PR: https://github.com/odoo/odoo/pull/165715

From: fa23f937934cb92cb9273b821bf9f2ce03aa2699
From: dhka-odoo
Date: 2024-06-14 14:30:59

Structural Changes: 6
Total Changes: 16

[IMP] hr: adding distance home-work unit field

adding new selection field distance home-work unit for changing between miles and km
added a new field for computation of miles into km
changing field name from km_home_work to distance_home_work
km_home_work will be used for calculation in km and distance_home_work will be used for input

task-3888505

closes odoo/odoo#165715

Related: odoo/enterprise#62600
Related: odoo/upgrade#6030
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr/models/hr_employee.py
+++ b/addons/hr/models/hr_employee.py
@@ -115,7 +115,12 @@ class HrEmployeePrivate(models.Model):
     study_school = fields.Char("School", groups="hr.group_hr_user", tracking=True)
     emergency_contact = fields.Char("Contact Name", groups="hr.group_hr_user", tracking=True)
     emergency_phone = fields.Char("Contact Phone", groups="hr.group_hr_user", tracking=True)
-    km_home_work = fields.Integer(string="Home-Work Distance", groups="hr.group_hr_user", tracking=True)
+    distance_home_work = fields.Integer(string="Home-Work Distance", groups="hr.group_hr_user", tracking=True)
+    km_home_work = fields.Integer(string="Home-Work Distance in Km", groups="hr.group_hr_user", compute="_compute_km_home_work", store=True)
+    distance_home_work_unit = fields.Selection([
+        ('kilometers', 'km'),
+        ('miles', 'mi'),
+    ], 'Home-Work Distance unit', tracking=True, groups="hr.group_hr_user", default='kilometers', required=True)
     employee_type = fields.Selection([
             ('employee', 'Employee'),
             ('student', 'Student'),
@@ -230,6 +235,11 @@ class HrEmployeePrivate(models.Model):
             permit_no = '_' + employee.permit_no if employee.permit_no else ''
             employee.work_permit_name = "%swork_permit%s" % (name, permit_no)
 
+    @api.depends('distance_home_work', 'distance_home_work_unit')
+    def _compute_km_home_work(self):
+        for employee in self:
+            employee.km_home_work = employee.distance_home_work * 1.609 if employee.distance_home_work_unit == "miles" else employee.distance_home_work
+
     def _get_partner_count_depends(self):
         return ['user_id']
 

--- a/addons/hr/models/res_users.py
+++ b/addons/hr/models/res_users.py
@@ -51,6 +51,8 @@ HR_WRITABLE_FIELDS = [
     'ssnid',
     'job_title',
     'km_home_work',
+    'distance_home_work',
+    'distance_home_work_unit',
     'marital',
     'mobile_phone',
     'employee_parent_id',
@@ -111,6 +113,8 @@ class User(models.Model):
     private_email = fields.Char(related='employee_id.private_email', string="Private Email", readonly=False)
     private_lang = fields.Selection(related='employee_id.lang', string="Employee Lang", readonly=False)
     km_home_work = fields.Integer(related='employee_id.km_home_work', readonly=False, related_sudo=False)
+    distance_home_work = fields.Integer(related='employee_id.distance_home_work', readonly=False, related_sudo=False)
+    distance_home_work_unit = fields.Selection(related='employee_id.distance_home_work_unit', readonly=False, related_sudo=False)
     # res.users already have a field bank_account_id and country_id from the res.partner inheritance: don't redefine them
     employee_bank_account_id = fields.Many2one(related='employee_id.bank_account_id', string="Employee's Bank Account Number", related_sudo=False, readonly=False)
     employee_country_id = fields.Many2one(related='employee_id.country_id', string="Employee's Country", readonly=False, related_sudo=False)
